### ISSUE TITLE
Adding autocomplete to cell editing

### ISSUE THREAD

**Issue by josephschorr** [glideapps/glide-data-grid#127]
First, I'd like to thank you for some a powerful component; it was quite easy to adopt and adapt.

We're making use of the data grid for table-like data entry, and one of the requests we've had is to provide autocompletion on cell editing if we have existing values matching the data being entered. Is there a known, good way to bind an autocompletion control to the cell editor provided by the data grid? If not, would it be possible to expose a ref or allow for a custom TextCell input component?

**Comment by JulianBertsch** [comment]
You could achieve it with a custom cell

**Comment by JulianBertsch** [comment]
@josephschorr  Not in documentation but I can share you a code example
`
// extraCellHook.js
import { useCallback } from 'react';
import { GridCellKind } from '@glideapps/glide-data-grid';
import TagsCellRenderer from '../customCells/tags-cell';
import DatePickerCellRenderer from '../customCells/datepicker-cell';
import StatusCellRenderer from '../customCells/status-cell';
import TextCellRenderer from '../customCells/text-cell';

const cells = [TextCellRenderer];

export function useExtraCells() {
	const drawCell = useCallback(args => {
		const {
			cell
		} = args;

		if (cell.kind !== GridCellKind.Custom && cell.kind !== GridCellKind.Text) return false;

		for (const c of cells) {
			if (c.isMatch(cell)) {
				return c.draw(args, cell);
			}
		}
		return false;
	}, []);

	const provideEditor = useCallback(cell => {

		if (cell.kind !== GridCellKind.Custom && cell.kind !== GridCellKind.Text) return undefined;

		for (const c of cells) {
			if (c.isMatch(cell)) {
				return c.provideEditor(cell);
			}
		}

		return cells[3].provideEditor(cell);
	}, []);

	return {
		drawCell,
		provideEditor
	};
}

// text-cell.js
import { useEffect, createRef } from 'react';
import { measureTextCached } from "@glideapps/glide-data-grid";
import { Form, Input, Select, Button } from 'antd';
import styled from 'styled-components';

const fontSize = 13;

const EditorWrap = styled.div`	
	transition: 2s ease-in;
	input {
		padding: 0px;
		&.ant-input {
			&:focus {
				box-shadow: rgb(0 0 0 / 20%) 0px 0px 0px 0px
			}
		}			
	}
`


const drawTextCell = (args, data) => {
	const {
		ctx,
		rect: {x, y, width: w, height: h},
		theme
	} = args;


	data = data.displayData;

	if (data.displayData) data=data.displayData;

	if(data) data = data.split(/\r?\n/)[0].slice(0, Math.round(w / 4));

	ctx.fillStyle = theme.textDark;
	ctx.font = `${fontSize}px ${theme.fontFamily}`;


	// const textHeight = measureTextCached(data, ctx, `${theme.baseFontStyle} ${theme.fontFamily}`);

	ctx.fillText(data, x + theme.cellHorizontalPadding + 0.5, y + h / 2 + fontSize / 2);

	return true;
}

const renderer = {
	isMatch: c => c.data.kind === "text-cell",
	draw: drawTextCell,
	provideEditor: () => {
		return p => {
			const {
				isHighlighted,
				onChange,
				onKeyDown,
				value
			} = p;

			let inputRef = null;

			const [form] = Form.useForm();

			const  getValue = () => {
				return value.data?.contentData?.contentData ?? value.data.contentData ?? value.data.data ?? value.data;
			}

			useEffect(() => {
				const stopScrollingOnContextMenu = (e) => {
					e.preventDefault();
				};
				window.addEventListener("mousewheel", stopScrollingOnContextMenu,  { passive: false });
				return () => {
					window.removeEventListener("mousewheel", stopScrollingOnContextMenu,  false);
				};
			}, []);

			useEffect(async () => {
				await form.setFieldsValue({field: getValue()});
				await form.validateFields();
				if (! form.getFieldError('field').length) onChange({
					...p.value,
					data: {
						...value.data,
						contentData: getValue(),
						displayData: getValue() ?? '-'
					},
				});
			}, []);

			useEffect(async () => {
				inputRef?.focus();
				// let length = getValue().length;
				// inputRef.setSelectionRange(isHighlighted ? length : 0, length);
			}, [inputRef]);

			const customRules = {
				phone: (errorMessage) => {
					return ({ getFieldValue }) => ({
						validator(_, value) {
							if (value.match(/^(?:(?:\(?(?:00|\+)([1-4]\d\d|[1-9]\d?)\)?)?[\-\.\ \\\/]?)?((?:\(?\d{1,}\)?[\-\.\ \\\/]?){0,})(?:[\-\.\ \\\/]?(?:#|ext\.?|extension|x)[\-\.\ \\\/]?(\d+))?$/i)) {
								return Promise.resolve();
							}
							return Promise.reject(new Error(errorMessage));
						},
					})
				}
			};

			const rules = value.rules ? value.rules.map(rule => {
				if (! rule.custom) return rule;
				return customRules[rule.type](rule.message)
			}) : [];

			// TODO: implement rules from column data
			return <EditorWrap>
			{value.readonly ? getValue() :<Form form={form}>
			<Form.Item style={{ width: '100%', height: '100%', margin: '0px' }} name="field" rules={rules}>
				 <Input
					autoComplete="off"
					type="textarea"
					defaultValue={getValue()}
					onChange={async e => {
						try {
							await form.validateFields();
							// prevent saving data if has errors
							if (! form.getFieldError('field').length) onChange({
								...p.value,
								data: {
									...value.data,
									contentData: e.target.value,
									displayData: e.target.value ? e.target.value : '-'
								},
							});
						} catch(e) {
							console.log('error', e);
						}

					}}
					ref={input => {
						inputRef = input;
					}}
					style={{ width: '100%' }}
				/>
			</Form.Item>
			</Form>}
			</EditorWrap>;
		}
	}
}

export default renderer;


// usage of the custom cells
const getData = useCallback(([col, row]) => {
  return {
			kind: GridCellKind.Text, // custom but this also triggers doFocus on new row appended,
			displayData: filteredDataSet[row][col] ? filteredDataSet[row][col] : '-',
			copyData: filteredDataSet[row][col] ? filteredDataSet[row][col] : '',
			rules: filteredCols[col].rules,
			data: {
				kind: 'text-cell',
				contentData: filteredDataSet[row][col] ? filteredDataSet[row][col] : '',
				copyData: filteredDataSet[row][col] ? filteredDataSet[row][col] : '',
				displayData: filteredDataSet[row][col] ? filteredDataSet[row][col] : '-'
			}
		};
});

`

**Comment by JulianBertsch** [comment]
@josephschorr  Not in documentation but I can share you a code example
`
// extraCellHook.js
import { useCallback } from 'react';
import { GridCellKind } from '@glideapps/glide-data-grid';
import TagsCellRenderer from '../customCells/tags-cell';
import DatePickerCellRenderer from '../customCells/datepicker-cell';
import StatusCellRenderer from '../customCells/status-cell';
import TextCellRenderer from '../customCells/text-cell';

const cells = [TextCellRenderer];

export function useExtraCells() {
	const drawCell = useCallback(args => {
		const {
			cell
		} = args;

		if (cell.kind !== GridCellKind.Custom && cell.kind !== GridCellKind.Text) return false;

		for (const c of cells) {
			if (c.isMatch(cell)) {
				return c.draw(args, cell);
			}
		}
		return false;
	}, []);

	const provideEditor = useCallback(cell => {

		if (cell.kind !== GridCellKind.Custom && cell.kind !== GridCellKind.Text) return undefined;

		for (const c of cells) {
			if (c.isMatch(cell)) {
				return c.provideEditor(cell);
			}
		}

		return cells[3].provideEditor(cell);
	}, []);

	return {
		drawCell,
		provideEditor
	};
}

// text-cell.js
import { useEffect, createRef } from 'react';
import { measureTextCached } from "@glideapps/glide-data-grid";
import { Form, Input, Select, Button } from 'antd';
import styled from 'styled-components';

const fontSize = 13;

const EditorWrap = styled.div`	
	transition: 2s ease-in;
	input {
		padding: 0px;
		&.ant-input {
			&:focus {
				box-shadow: rgb(0 0 0 / 20%) 0px 0px 0px 0px
			}
		}			
	}
`


const drawTextCell = (args, data) => {
	const {
		ctx,
		rect: {x, y, width: w, height: h},
		theme
	} = args;


	data = data.displayData;

	if (data.displayData) data=data.displayData;

	if(data) data = data.split(/\r?\n/)[0].slice(0, Math.round(w / 4));

	ctx.fillStyle = theme.textDark;
	ctx.font = `${fontSize}px ${theme.fontFamily}`;


	// const textHeight = measureTextCached(data, ctx, `${theme.baseFontStyle} ${theme.fontFamily}`);

	ctx.fillText(data, x + theme.cellHorizontalPadding + 0.5, y + h / 2 + fontSize / 2);

	return true;
}

const renderer = {
	isMatch: c => c.data.kind === "text-cell",
	draw: drawTextCell,
	provideEditor: () => {
		return p => {
			const {
				isHighlighted,
				onChange,
				onKeyDown,
				value
			} = p;

			let inputRef = null;

			const [form] = Form.useForm();

			const  getValue = () => {
				return value.data?.contentData?.contentData ?? value.data.contentData ?? value.data.data ?? value.data;
			}

			useEffect(() => {
				const stopScrollingOnContextMenu = (e) => {
					e.preventDefault();
				};
				window.addEventListener("mousewheel", stopScrollingOnContextMenu,  { passive: false });
				return () => {
					window.removeEventListener("mousewheel", stopScrollingOnContextMenu,  false);
				};
			}, []);

			useEffect(async () => {
				await form.setFieldsValue({field: getValue()});
				await form.validateFields();
				if (! form.getFieldError('field').length) onChange({
					...p.value,
					data: {
						...value.data,
						contentData: getValue(),
						displayData: getValue() ?? '-'
					},
				});
			}, []);

			useEffect(async () => {
				inputRef?.focus();
				// let length = getValue().length;
				// inputRef.setSelectionRange(isHighlighted ? length : 0, length);
			}, [inputRef]);

			const customRules = {
				phone: (errorMessage) => {
					return ({ getFieldValue }) => ({
						validator(_, value) {
							if (value.match(/^(?:(?:\(?(?:00|\+)([1-4]\d\d|[1-9]\d?)\)?)?[\-\.\ \\\/]?)?((?:\(?\d{1,}\)?[\-\.\ \\\/]?){0,})(?:[\-\.\ \\\/]?(?:#|ext\.?|extension|x)[\-\.\ \\\/]?(\d+))?$/i)) {
								return Promise.resolve();
							}
							return Promise.reject(new Error(errorMessage));
						},
					})
				}
			};

			const rules = value.rules ? value.rules.map(rule => {
				if (! rule.custom) return rule;
				return customRules[rule.type](rule.message)
			}) : [];

			// TODO: implement rules from column data
			return <EditorWrap>
			{value.readonly ? getValue() :<Form form={form}>
			<Form.Item style={{ width: '100%', height: '100%', margin: '0px' }} name="field" rules={rules}>
				 <Input
					autoComplete="off"
					type="textarea"
					defaultValue={getValue()}
					onChange={async e => {
						try {
							await form.validateFields();
							// prevent saving data if has errors
							if (! form.getFieldError('field').length) onChange({
								...p.value,
								data: {
									...value.data,
									contentData: e.target.value,
									displayData: e.target.value ? e.target.value : '-'
								},
							});
						} catch(e) {
							console.log('error', e);
						}

					}}
					ref={input => {
						inputRef = input;
					}}
					style={{ width: '100%' }}
				/>
			</Form.Item>
			</Form>}
			</EditorWrap>;
		}
	}
}

export default renderer;


// usage of the custom cells
const getData = useCallback(([col, row]) => {
  return {
			kind: GridCellKind.Text, // custom but this also triggers doFocus on new row appended,
			displayData: filteredDataSet[row][col] ? filteredDataSet[row][col] : '-',
			copyData: filteredDataSet[row][col] ? filteredDataSet[row][col] : '',
			rules: filteredCols[col].rules,
			data: {
				kind: 'text-cell',
				contentData: filteredDataSet[row][col] ? filteredDataSet[row][col] : '',
				copyData: filteredDataSet[row][col] ? filteredDataSet[row][col] : '',
				displayData: filteredDataSet[row][col] ? filteredDataSet[row][col] : '-'
			}
		};
});

`

**Comment by JulianBertsch** [comment]
@josephschorr You also have to implement the custom cell drawing function and the custom editors on the DataEditor component with the properties "drawCell={drawCell}
						provideEditor={provideEditor}"

**Comment by josephschorr** [comment]
@JulianBertsch Thanks!

**Comment by josephschorr** [comment]
@JulianBertsch Thanks for the code sample. I initially tried to use it, but ran into the issue that it has `kind: GridCellKind.Text`, but a `data` field of a non-string value. I eventually got the autocompletion working with pure custom cells, but now I've lost the nice abilities associated with a text cell, such as delete-clearing, hitting any key opening the editor, etc.

Is there a way to have a custom cell type, but retain (or implement) these capabilities? If not, I'll go back to cells of type string, with inference based on the column position, but I'd prefer if I could use custom cells.

**Comment by jassmith** [comment]
I dont understand, can you fork the storybook and make a story that shows your issue with editing the editor for the text cell?

**Comment by josephschorr** [comment]
@jassmith Sure, I can try to put that together tomorrow. In the meantime, perhaps I can explain it in a more coherent fashion: I initially tried to use the example @JulianBertsch put together for custom cells, but this failed to type check:

```
return {
  kind: GridCellKind.Text,
  data: {
    kind: 'text-cell',
  },
  ...
}
```

because `kind` being `GridCellKind.Text` must have `data` of type `string`. So instead, I switched it to `kind: GridCellKind.Custom`.

Unfortunately, this disables all the nice "default" edit operations, such as the ability to select a cell and hit the `DELETE` key to clear it, or to have the editor immediately appear if you hit an alphanumeric key on the keyboard while having a cell selected.

Thus, I'm torn on what to do: should I change my cell type back to `GridCellKind.Text`, and have a single custom `provideEditor`, or is there a way we can extend the control to allow me to mark my `GridCellKind.Custom` cells as being "editable", such that I can get similar behavior to a text cell?

Ideally, I can keep my cells as custom and enable this extra behavior, but I don't believe there are hooks yet to customize that within the control.

Apologies if this is still unclear :)

**Comment by jassmith** [comment]
I see, so to be extra clear, would your problem be addressed if:

- Custom cells with custom editors could be triggered on alphanumeric press
- Delete key could be handled by custom editors as well

**Comment by josephschorr** [comment]
Yep!

**Comment by JulianBertsch** [comment]
@jassmith @josephschorr  the reason why I gave the Gridcell.text but with customfield values ist to that the package code triggers the custom "overwritten" textfield. There you have to conditioning for the textfield into your custom fields hook and return in case of textfield the customfield. That's just a workaround because we can not yet define which field it should open automatically when you press key or adding new row, which would solve the issue

**Comment by josephschorr** [comment]
@jassmith No rush! I was just pointing it out in case anyone else got confused when trying in TS as well.

In terms of PR, I'd be happy to write one, if you can point me to the existing code for those actions.

**Comment by jassmith** [comment]
Those handlers are both in data-editor.tsx `onKeyDown`. The easiest way to figure this out is to go sign up for replay.io and make yourself a recording of performing both interactings using a custom and a standard editor. Then you can just trace the code path out. Honestly this project would not be possible without using replay. The drawing code would be just impossible to get right.

**Comment by jassmith** [comment]
This is fixed in 3.5.0. It is now possible to respond to initial values and delete events with custom cells
